# Explanation of the task

We have 1418 and 1912 images of men and women respectively. Our task is to create a deep learning classification algorithm under different optimizers and paramters using Keras.

In [1]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
from PIL import Image
import numpy as np
import os
import cv2
import keras
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Conv2D,MaxPooling2D,Dense,Flatten,Dropout

Using TensorFlow backend.


In [2]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['data']


# Data Preparation
Data Preperation: We will make data and labels list where data will be image to array implementatation which contains RGB values of each image. and label will be class of cells here I will be taking 0 and 1 for two classes

In [3]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
data=[]
labels=[]
men=os.listdir("../input/data/men/")
dummy = 0
for a in men:
    try:
        image=cv2.imread("../input/data/men/"+a)
        image_from_array = Image.fromarray(image, 'RGB')
        size_image = image_from_array.resize((200, 200))
        data.append(np.array(size_image))
        labels.append(0)
    except AttributeError:
        dummy += 1
women=os.listdir("../input/data/women/")
for b in women:
    try:
        image=cv2.imread("../input/data/women/"+b)
        image_from_array = Image.fromarray(image, 'RGB')
        size_image = image_from_array.resize((200, 200))
        data.append(np.array(size_image))
        labels.append(1)
    except AttributeError:
        dummy += 1

In [4]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
Cells=np.array(data)
labels=np.array(labels)
print(Cells.shape)
print(labels.shape)

(3311, 200, 200, 3)
(3311,)


### Meaning of Cells.shape and labels.shape

Cells.shape has 3311 data points (ie images). Each data point, is represented by three 200 by 200 matrices, because we have one matrix for each RGB colour.

labels.shape has 3311 labels for each image.

In [5]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
np.save("Cells",Cells)
np.save("labels",labels)

In [6]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
Cells=np.load("Cells.npy")
labels=np.load("labels.npy")

In [7]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
s=np.arange(Cells.shape[0])
np.random.shuffle(s)
Cells=Cells[s]
labels=labels[s]

In [8]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
num_classes=len(np.unique(labels))
len_data=len(Cells)

print(num_classes)
print(len_data)

2
3311


### Meaning of num_classes and len_data

num_classes is the number of classes, len_data is the number of images.

In [9]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
(x_train,x_test)=Cells[(int)(0.1*len_data):],Cells[:(int)(0.1*len_data)]
x_train = x_train.astype('float32')/255 # As we are working on image data we are normalizing data by dividing 255.
x_test = x_test.astype('float32')/255
train_len=len(x_train)
test_len=len(x_test)

In [10]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
(y_train,y_test)=labels[(int)(0.1*len_data):],labels[:(int)(0.1*len_data)]
print(x_train.shape)
print(y_train.shape)

(2980, 200, 200, 3)
(2980,)


### Meaning of x_train.shape and y_train.shape

x_train has the data for 2980 images and y_train has the labels for them.

# Modeling with Keras Sequential
## (with different optimizers)

We will use the categorical cross entropy as the loss function.
The categorical cross entropy (aka log-loss) penalizes the accuracy of a probability prediction.
For example, if an algorithm predicts a probability of 0.6 for an image with label 1, the penalty for this prediction will be log(.6). If the label was 0, the penalty will be log(0.4).

In [11]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
#creating sequential model
model=Sequential()
model.add(Conv2D(filters=16,kernel_size=2,padding="same",activation="relu",input_shape=(200,200,3)))
model.add(MaxPooling2D(pool_size=2))
model.add(Conv2D(filters=32,kernel_size=2,padding="same",activation="relu"))
model.add(MaxPooling2D(pool_size=2))
model.add(Conv2D(filters=64,kernel_size=2,padding="same",activation="relu"))
model.add(MaxPooling2D(pool_size=2))
model.add(Dropout(0.2))
model.add(Flatten())
model.add(Dense(500,activation="relu"))
model.add(Dropout(0.2))
model.add(Dense(2,activation="softmax"))#2 represent output layer neurons 
model.summary()

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 200, 200, 16)      208       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 100, 100, 16)      0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 100, 100, 32)      2080      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 50, 50, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 50, 50, 64)        8256      
_________________________________________________________________
max_pooling2d_3 (MaxP

### Intuitive meaning of the sequential model defined above

There are three 2-dimensional convolutional filters.
The filter sizes are (in order) 16, 32 and 64.
For example, this means the last convolutional filter reduces the input layer more substantially than the second filter because the neuron looks at 32 pixels at once on the second filter but 64 pixels on the last filter.

Each 2-dimensional convolutional filter is followed by a max pooling operation for 2-dimensions.
The max pooling operation performs dimension reduction on the output layer from the previous convolutional filter before it becomes the output layer for the next convolutional filter.

Dropout's purpose is to avoid overfitting. It aims to achieve this by randomly selecting elements from the matrix and replacing them with 0.

Flatten means transforming the feature matrix into a vector.

Dense represents a densley-connected neural network layer which performs a linear operation on a vector.
The first dc-nn layer's output vector is 500-dimensional and the second is 2-dimensional.
The two activations for Dense used above are relu and softmax.

# Adam Optimizer with default parameters

Objective: 
Minimize the loss function with parameter theta.

Parameters: 
Step size denoted by alpha.
Decay rate for first moment estimate denoted by beta1.
Decay rate for second raw moment estimate denoted by beta2.
Initial paremeter vector for theta. 

Intuitive meaning of algorithm steps: 

(1) Get the gradients of the loss function under the parameter vector for theta. 

(2) Update the first moment estimates by: beta1 x previous iteration moment estimates + (1-beta1) x gradients of loss function. 

(3) Update the second raw moment estimates by: beta1 x previous iteration moment estimates + (1-beta1) x squared gradients of loss function. 
Notice the larger beta is, the more bias and less variant the estimates above will be and vice-versa.

(4) Bias-correct the first moment estimates by: current estimate / (1-beta1). 

(5) Bias correct the second raw moment estimates by: current estimate / (1-beta2). 
Notice the larger beta is, the larger the bias-corrected estimates will be and thus moving further away from convergence; vice-versa.

(6) Update the parameter vector by: previous parameter vector - alpha x first moment estimates / sqrt of second moment estimates.

Iterate until the updated parameter vector are essentially equal to the previous iteration's parameter vector.

Reference (Kingma&Ba, 2015): https://arxiv.org/pdf/1412.6980v8.pdf

In [12]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
#Doing One hot encoding as classifier has multiple classes
y_train=keras.utils.to_categorical(y_train,2)
y_test=keras.utils.to_categorical(y_test,2)

### Step 1: Compile

In [13]:
#Source code: https://www.kaggle.com/tjm2018/starter-men-women-classification-e6719600-3
#compile the model with loss as categorical_crossentropy and using adam optimizer
model.compile(loss='categorical_crossentropy', optimizer='adam')

### Step 2: Fit

In [14]:
model.fit(x_train, y_train)

Instructions for updating:
Use tf.cast instead.
Epoch 1/1
2980/2980 [==============================] - 6s 2ms/step - loss: 0.7807


### Step 3: Evaluate

In [15]:
model.evaluate(x_train, y_train)

2980/2980 [==============================] - 1s 424us/step


0.6417159494137604

### Conclusion for Adam Optimizer default:

It's empirical loss (as shown on step 2's output) is much greater than the true loss (as shown on step 3's output). This raises sucipicioun of an underfit. Therefore we will re-train the model with smaller step sizes (i.e. decrease alpha).

# Adam Optimizer with small alpha

The default aplpha value is number of samples in dataset divided by 32. We will change this denominator to 64 to halve the alpha value.

In [16]:
model.fit(x_train, y_train, batch_size = 64)
model.evaluate(x_train, y_train)

Epoch 1/1
2980/2980 [==============================] - 1s 395us/step


0.6047906748400439

Success. We realise the empirical loss is still larger than the true loss but the discrepancy has decreased significantly. More importantly, the true loss has decreased. Now we will observe the Adamax Optimizer which was introduced by the same paper that introduced Adam Optimizer.

# Adamax Optimizer

Objective: 
Minimize the loss function with parameter theta.

Parameters: 
Step size denoted by alpha.
Decay rate for first moment estimate denoted by beta1.
Decay rate for exponentially wieghted inifinity norm denoted by beta2.
Initial paremeter vector for theta. 

Intuitive meaning of algorithm steps: 

(1) Get the gradients of the loss function under the parameter vector for theta.

(2) Update the first moment vector by: beta1 x current first moment vector + (1 - beta1) x gradients of the loss function

(3) Update the exponentially wieghted inifinity norm by: max( b2 x exponentially wieghted inifinity norm, absolute value gradients of the loss function )

(4) Update parameter vector by: current parameter vector - (alpha / (1 - beta1^iteration_count) ) * first moment vector / exponentially wieghted inifinity norm

Iterate until the updated parameter vector are essentially equal to the previous iteration's parameter vector.

Reference (Kingma&Ba, 2015): https://arxiv.org/pdf/1412.6980v8.pdf

Now we will fit this optimizer for our dataset with default parameters and then by halving alpha like we did for observing Adam.

In [17]:
model.compile(loss='categorical_crossentropy', optimizer='adamax')
model.fit(x_train, y_train)
model.evaluate(x_train, y_train)

Epoch 1/1
2980/2980 [==============================] - 1s 411us/step


0.5420882996296723

In [18]:
model.compile(loss='categorical_crossentropy', optimizer='adamax')
model.fit(x_train, y_train, batch_size = 64)
model.evaluate(x_train, y_train)

Epoch 1/1
2980/2980 [==============================] - 1s 428us/step


0.48027771699348554

For Adamax, we notice the discrepancy between the true loss and empirical loss doesn't change a lot for the different alpha values, which is unlike what we observed for Adam. However, again, we notice that the true loss has decreased after alpha was halved.

When we compare the performance of Adam and Adamax for this dataset, Adam seems to perform better in terms of over/under fitting. This supports what was mentioned on the paper, how Adamax suits loss functions L3-norm or higher. Our loss function, categorical crossentropy function, is log L1-norm.